jctools-core/src/main/java/org/jctools/queues/MpscProgressiveChunkedQueue.java

```diff
@@ -0,0 +1,553 @@
+        final int pOffset = (int) (producerSeq & chunkMask);
+        final long chunkIndex = producerSeq >> chunkShift;
+        final boolean firstElementOfNewChunk = pOffset == 0 && chunkIndex > 0;
+        final long currentChunkIndex = firstElementOfNewChunk ? chunkIndex - 1 : chunkIndex;
+        final AtomicChunk<E> producerBuffer = producerBufferOf(null, currentChunkIndex);
+        if (firstElementOfNewChunk)
+        {
+            appendElementAndLinkChunk(producerBuffer, chunkIndex, chunkMask + 1, e, null);
+        }
+        else
+        {
+            producerBuffer.soElement(pOffset, e);
+        }
+        return true;
+    }
+
+    private void appendElementAndLinkChunk(
+        AtomicChunk<E> producerBuffer,
+        long chunkIndex,
+        int chunkSize,
+        E e,
+        Supplier<E> s)
+    {
+        //we can create the new chunk
+        AtomicChunk<E> newChunk = lvFreeBuffer();
+        if (newChunk != null)
+        {
+            if (casFreeBuffer(newChunk, null))
+            {
+                newChunk.init(chunkIndex, producerBuffer);
+            }
+            else
+            {
+                newChunk = null;
+            }
+        }
+        if (newChunk == null)
+        {
+            newChunk = new AtomicChunk<E>(chunkIndex, producerBuffer, chunkSize);
+        }
+        if (e == null)
+        {
+            e = s.get();
+        }
+        newChunk.soElement(0, e);
+        producerBuffer.soNext(newChunk);
+        soProducerBuffer(newChunk);
+    }
+
+
+    private static <E> E spinForElement(AtomicChunk<E> chunk, int offset)
+    {
+        E e;
+        while ((e = chunk.lvElement(offset)) == null)
+        {
+
+        }
+        return e;
+    }
+
+    @Override
+    public E poll()
+    {
+        final long chunkMask = this.chunkMask;
+        final long consumerIndex = this.lpConsumerIndex();
+        AtomicChunk<E> consumerBuffer = this.consumerBuffer;
+        final int consumerOffset = (int) (consumerIndex & chunkMask);
+        if (consumerOffset == 0 && consumerIndex >= (chunkMask + 1))
+        {
+            AtomicChunk<E> next = consumerBuffer.lvNext();
+            if (next == null)
+            {
+                if (lvProducerIndex() == consumerIndex)
+                {
+                    return null;
+                }
+                while ((next = consumerBuffer.lvNext()) == null)
+                {
+
+                }
+            }
+            //save from nepotism: throw away consumerBuffer
+            consumerBuffer.spNext(null);
+            //recycle buffer if necessary: optimistic reciclying, given
+            //that a racing producer can already obtain it
+            if (lpFreeBuffer() == null)
+            {
+                soFreeBuffer(consumerBuffer);
+            }
+            next.spPrev(null);
+            consumerBuffer = next;
+            this.consumerBuffer = consumerBuffer;
+        }
+        E e = consumerBuffer.lvElement(consumerOffset);
```

Comment:
The following code can be re-arranged as elsewhere for single consumer code

Reply:
Do you mean by grouping everything under a single method or just by grouping the "slow" path:
```java
        assert !firstElementOfNewChunk;
        if (lvProducerIndex() == consumerIndex)
        {
            return null;
        }
        e = spinForElement(consumerBuffer, consumerOffset);
        consumerBuffer.spElement(consumerOffset, null);
        soConsumerIndex(consumerIndex + 1);
        return e;
```